portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -49,168 +49,203 @@
-  
-  private void putBlogsInDatastore(String tag, String message, String nickname, List<BlogMessage> reply, long parentID) {
-    Entity blogMessageEntity = new Entity(BLOG_ENTITY_KIND);
-    blogMessageEntity.setProperty("nickname", nickname);
-    blogMessageEntity.setProperty("text", message);
-    blogMessageEntity.setProperty("time", System.currentTimeMillis());
-    blogMessageEntity.setProperty("tag", tag);
-    blogMessageEntity.setProperty("replies", reply);
-    blogMessageEntity.setProperty("parentID", parentID);
-    
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    datastore.put(blogMessageEntity);
-  }
-
-  private List<BlogMessage> getBlogsFromDatastore() {
-    List<BlogMessage> BlogMessages = new ArrayList<BlogMessage>();
-    Query query = new Query(BLOG_ENTITY_KIND).addSort("time", SortDirection.ASCENDING);
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    PreparedQuery results = datastore.prepare(query);
-
-    UserService userService = UserServiceFactory.getUserService();
-
-    for (Entity entity : results.asIterable()) {
-      long messageId = entity.getKey().getId();
-      long timestamp = (long) entity.getProperty("time");
-      String tag = (String) entity.getProperty("tag");
-      String comment = (String) entity.getProperty("text");
-      String nickname = (String) entity.getProperty("nickname");
-      String email = (String) userService.getCurrentUser().getEmail();
-      long parentID = (long) 0;
-      // TODO: UPDATE WHEN FRONTEND INPUT IS TAKEN 
-      ArrayList<BlogMessage> messageReplies = (ArrayList) entity.getProperty("replies");
-      BlogMessage message = new BlogMessage(
-            messageId, tag, comment, nickname, email, messageReplies, timestamp, parentID);
-      BlogMessages.add(message);
-    }
-    return BlogMessages;
-  }
-    
+ 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get BlogMessages from Datastore.
-    List<BlogMessage> BlogMessages = getBlogsFromDatastore();
-
-    // Separate posts from replies from BlogMessages. 
+    List<BlogMessage> blogMessages = LoadAllBlogsOrLast(true);
+ 
+    // TODO: Get these from client.
+    int numberOfCommentsToDisplay = 0;
+    List<String> tagsToSearch = new ArrayList<String>();
+ 
+    // TODO: Use helper functions for lines 73-101.
+ 
+    // Separate posts from replies from |blogMessages|.
     // Add replies to messageReplies for the respective posts.
-    // Create BlogHashMap Object and put BlogMessages in the map.
     List<BlogMessage> blogMessagesReplies = new ArrayList<BlogMessage>();
-    for (BlogMessage message : BlogMessages) {
+    for (BlogMessage message : blogMessages) {
       if (message.getParentID() != 0) {
         blogMessagesReplies.add(message);
-        BlogMessages.remove(message);
+        blogMessages.remove(message);
```

Comment:
You probably cannot remove an element during iterating over the same container. Some reference[1]. It may mess up with the underlying iterator and throw exceptions.

[1] https://stackoverflow.com/questions/10431981/remove-elements-from-collection-while-iterating